python/rmm/_cuda/gpu.py

```diff
@@ -1,59 +1,43 @@
 # Copyright (c) 2020, NVIDIA CORPORATION.
 
-from rmm._cuda.gpu cimport (
-    CUresult,
-    cudaDeviceAttr,
-    cudaDeviceGetAttribute,
-    cudaDeviceProp,
-    cudaDriverGetVersion,
-    cudaError,
-    cudaError_t,
-    cudaGetDeviceCount,
-    cudaGetDeviceProperties,
-    cudaGetErrorName,
-    cudaGetErrorString,
-    cudaRuntimeGetVersion,
-    cuDeviceGetName,
-    cuGetErrorName,
-    cuGetErrorString,
-)
-
-from enum import IntEnum
+from cuda import cuda, cudart
 
 
 class CUDARuntimeError(RuntimeError):
-
-    def __init__(self, cudaError_t status):
+    def __init__(self, status: cudart.cudaError_t):
         self.status = status
-        cdef str name = cudaGetErrorName(status).decode()
-        cdef str msg = cudaGetErrorString(status).decode()
+
+        err, name = cudart.cudaGetErrorName(status)
+        if err != cudart.cudaError_t.cudaSuccess:
+            raise CUDARuntimeError(err.value)
```

Comment:
It's awkward that we might raise a `CUDARuntimeError` from within the constructor of `CUDARuntimeError`... which got me to looking a little closer.

I saw the suggestion from @jakirkham in https://github.com/rapidsai/rmm/pull/930#discussion_r782475266 / 8be6c0b but I'm pretty sure this is a spurious error check that illuminates some slightly awkward design in cuda-python (a meaningless `err` value).

For `cudaGetErrorName`, the `err` value is hardcoded as `cudaError_t.cudaSuccess`. It's not likely that this will ever change, because the corresponding runtime API does not generate an CUDA error: [cudaGetErrorName](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__ERROR.html#group__CUDART__ERROR_1gb3de7da2f23736878270026dcfc70075) returns a `const char*` with a special value of `"unrecognized error code"` rather than setting an error code if the parameters are invalid.
https://github.com/NVIDIA/cuda-python/blob/746b773c91e1ede708fe9a584b8cdb1c0f32b51d/cuda/cudart.pyx#L8084

Similarly for `cudaGetErrorString`... sort of:
https://github.com/NVIDIA/cuda-python/blob/746b773c91e1ede708fe9a584b8cdb1c0f32b51d/cuda/cudart.pyx#L8115
This function actually _can_ set an error code via its call to the driver API `cuGetErrorString`, but that error is not returned -- the `cudaGetErrorString` in cuda-python has `err` hard-coded as a success and the value `"unrecognized error code"` is returned. https://github.com/NVIDIA/cuda-python/blob/746b773c91e1ede708fe9a584b8cdb1c0f32b51d/cuda/_lib/ccudart/ccudart.pyx#L522-L531

In summary, I think it would be safe to revert 8be6c0b and use the previous snippet with a throwaway variable `_, name = ...` for both `cudaGetErrorName` and `cudaGetErrorString` because the `err` is always a success. As a consequence, we don't need to handle this awkward possibility of raising an error class in its own constructor.

Reply:
I agree that it is awkward. Would hope this error can't come up, but it might come if the error code is new. IOW only shows up in later CUDA versions, but not earlier ones. Or it could show up if the error code is malformed.

Maybe a middle ground would be to raise a `RuntimeError` so we don't have a recursive `CUDARuntimeError` definition. Thoughts?

Reply:
If a new or malformed error code is provided, the returned string from `cudaGetErrorName` or `cudaGetErrorString` will say `"unrecognized error code"` but no CUDA error will be set. Thus we do not need to check for a CUDA error from those functions. The cuda-python decision to hard-code the `err` value as "success" strikes me as odd because no `err` should be returned _at all_.

If we were to need to raise an error, the recursive use of `CUDARuntimeError` is not problematic/incorrect -- just awkward.

However, I don't have strong enough feelings on this to hold back on merging.

Reply:
Perhaps we should check for that build string and then error? Having suppressed errors doesn't sound good.

Agree with your assessment CUDA-Python should be handling this better, but maybe we can have that discussion in a new CUDA-Python issue?

Reply:
That sounds like a good idea! Raise a RuntimeError if we get the unrecognized error code string back, but don’t check the value of `err`.

Reply:
Fix'd!

Reply:
After a brief sync offline with Bradlet, we decided to just not worry about handling an unrecognized error code, because we couldn't figure out how such an error code could be constructed to begin with. `cudaError_t` is an `enum` that cannot be constructed from arbitrary integers.